packages/graphql/src/lib/schema.ts

```diff
@@ -0,0 +1,105 @@
+import {
+  type DecoratorContext,
+  type DecoratorFunction,
+  type Namespace,
+  type Program,
+  type Type,
+  getTypeName,
+  validateDecoratorUniqueOnNode,
+} from "@typespec/compiler";
+
+import { createStateSymbol, NAMESPACE, reportDiagnostic } from "../lib.js";
+import { useStateMap } from "./state-map.js";
+
+// This will set the namespace for decorators implemented in this file
+export const namespace = NAMESPACE;
+
+export interface SchemaDetails {
+  name?: string;
+}
+
+export interface Schema extends SchemaDetails {
+  type: Namespace;
+}
+
+const [getSchema, setSchema, getSchemaMap] = useStateMap<Namespace, Schema>(
+  createStateSymbol("schemas"),
+);
+
+/**
+ * List all the schemas defined in the TypeSpec program
+ * @param program Program
+ * @returns List of schemas.
+ */
+export function listSchemas(program: Program): Schema[] {
+  return [...getSchemaMap(program).values()];
+}
+
+export {
+  /**
+   * Get the schema information for the given namespace.
+   * @param program Program
+   * @param namespace Schema namespace
+   * @returns Schema information or undefined if namespace is not a schema namespace.
+   */
+  getSchema,
+};
+
+/**
+ * Check if the namespace is defined as a schema.
+ * @param program Program
+ * @param namespace Namespace
+ * @returns Boolean
+ */
+export function isSchema(program: Program, namespace: Namespace): boolean {
+  return getSchemaMap(program).has(namespace);
+}
+
+/**
+ * Mark the given namespace as a schema.
+ * @param program Program
+ * @param namespace Namespace
+ * @param details Schema details
+ */
+export function addSchema(
+  program: Program,
+  namespace: Namespace,
+  details: SchemaDetails = {},
+): void {
+  const schemaMap = getSchemaMap(program);
+  const existing = schemaMap.get(namespace) ?? {};
+  setSchema(program, namespace, { ...existing, ...details, type: namespace });
+}
+
+export const $schema: DecoratorFunction = (
+  context: DecoratorContext,
+  target: Namespace,
+  options: Type,
+) => {
+  validateDecoratorUniqueOnNode(context, target, $schema);
+
+  if (options && options.kind !== "Model") {
```

Comment:
Does this need to be a specific type of model i.e. `SchemaOptions`?

Reply:
Interesting question. I followed the pattern of the `@service` decorator here, but it seems that [in more recent work](https://github.com/timotheeguerin/typespec/blob/3653a3591f7401b0a2f21578061bdd01d164c43f/.chronus/changes/uri-templates-2024-6-25-9-3-39.md) there's a preference to use an object literal rather than a model for such options: 
> Decorator is also expecting an object value now instead of a model.

I'm not sure why; both appear to validate the input equivalently. But the nice thing about the object value approach is that we can annotate the type in the TypeScript implementation, so I'll go with that.